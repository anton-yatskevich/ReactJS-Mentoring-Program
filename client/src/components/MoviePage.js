import React from 'react';
import PropTypes from 'prop-types';

import Movie from './Movie';
import ResultsList from './ResultsList';

const MoviePage = ({ movies, selectedMovie, onSelectMovie }) => {
    const currentGenre = selectedMovie.genres[0];
    const fiteredResults = movies
        .filter(movie => movie.genres.some(genre => currentGenre === genre));

    return (
        <>
            <Movie {...selectedMovie} />
            <div className="search-description__wrapper">
                <p className="movie-page__results-description">{`Films by ${currentGenre} genre`}</p>
            </div>
            <ResultsList movies={fiteredResults} onSelectMovieHandler={onSelectMovie} />
        </>
    );
};

const movieSchema = PropTypes.objectOf(
    PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(PropTypes.string)
    ])
);

MoviePage.propTypes = {
    movies: PropTypes.arrayOf(movieSchema),
    selectedMovie: movieSchema,
    onSelectMovie: PropTypes.func
};

MoviePage.defaultProps = {
    movies: [],
    selectedMovie: {},
    onSelectMovie: null
};

export default MoviePage;
