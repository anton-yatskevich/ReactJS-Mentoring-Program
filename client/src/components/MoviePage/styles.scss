.movie-page {
	&__results-description {
		max-width: 1200px;
		margin: 0 auto;
		font-size: 18px;
		font-weight: bold;
	}
}