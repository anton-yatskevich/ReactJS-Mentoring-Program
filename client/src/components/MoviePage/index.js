import React from 'react';
import PropTypes from 'prop-types';

import Movie from '../Movie';
import ResultsList from '../../containers/ResultsListContainer';
import MovieSchema from '../../constants/MovieSchema';
import './styles.scss';

const MoviePage = ({ movies, selectedMovie }) => {
    const currentGenre = selectedMovie.genres && selectedMovie.genres[0];
    const fiteredResults = movies
        .filter(movie => movie.genres.some(genre => currentGenre === genre));

    return (
        <>
            <Movie {...selectedMovie} />
            <div className="search-description__wrapper">
                <p className="movie-page__results-description">{`Films by ${currentGenre} genre`}</p>
            </div>
            <ResultsList movies={fiteredResults} />
        </>
    );
};

MoviePage.propTypes = {
    movies: PropTypes.arrayOf(MovieSchema),
    selectedMovie: MovieSchema
};

MoviePage.defaultProps = {
    movies: [],
    selectedMovie: {}
};

export default MoviePage;
