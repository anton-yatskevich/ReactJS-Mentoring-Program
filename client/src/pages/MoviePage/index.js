import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import noop from 'lodash.noop';
import MovieSchema from '../../constants/MovieSchema';
import { getSelectedMovie } from '../../actions/selectMovie';
import fetchMovies from '../../actions/fetchMovies';
import setSelectedMovieId from '../../actions/setSelectedMovieId';
import ResultsList from '../../containers/ResultsListContainer';
import Movie from '../../components/Movie';
import './styles.scss';

class MoviePage extends Component {
    static async getInitialProps({ store, query: { id } }) {
        store.dispatch(setSelectedMovieId(id));
        await store.dispatch(getSelectedMovie());
        await store.dispatch(fetchMovies());
    }

    componentDidMount() {
        const {
            router: { query: { id } },
            fetchSelectedMovie,
            setMovieId,
            getMovies
        } = this.props;
        if (id) {
            setMovieId(id);
            fetchSelectedMovie();
            getMovies();
        }
    }

    render() {
        const { selectedMovie } = this.props;

        return (
            <>
                {
                    selectedMovie && (
                        <>
                            <Movie {...selectedMovie} />
                            {
                                selectedMovie.genres && (
                                    <div className="search-description__wrapper">
                                        <p className="movie-page__results-description">{`Films by ${selectedMovie.genres[0]} genre`}</p>
                                    </div>
                                )
                            }
                        </>
                    )
                }
                <ResultsList />
            </>
        );
    }
}

MoviePage.propTypes = {
    selectedMovie: MovieSchema,
    fetchSelectedMovie: PropTypes.func,
    setMovieId: PropTypes.func,
    getMovies: PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    router: PropTypes.object.isRequired
};

MoviePage.defaultProps = {
    selectedMovie: {},
    fetchSelectedMovie: noop,
    setMovieId: noop,
    getMovies: noop
};

function mapStateToProps({ movies }) {
    return {
        selectedMovie: movies.selectedMovie
    };
}

export default connect(
    mapStateToProps,
    { fetchSelectedMovie: getSelectedMovie, setMovieId: setSelectedMovieId, getMovies: fetchMovies }
)(MoviePage);
