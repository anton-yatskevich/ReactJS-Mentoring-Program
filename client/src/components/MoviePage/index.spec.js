import React from 'react';
import { shallow } from 'enzyme';
import MoviePage from './index';

describe('MoviePage component', () => {
    it('should be rendered with default props', () => {
        const MoviePageComponent = shallow(<MoviePage />);
        expect(MoviePageComponent).toMatchSnapshot();
    });

    it('should be rendered with filtered results', () => {
        const props = {
            movies: [{
                genres: ['a', 'b']
            }],
            selectedMovie: {
                genres: ['a']
            }
        };
        const MoviePageComponent = shallow(<MoviePage {...props} />);
        expect(MoviePageComponent).toMatchSnapshot();
    });
});
