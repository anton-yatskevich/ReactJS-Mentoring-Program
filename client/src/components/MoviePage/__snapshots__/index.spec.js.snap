// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage component should be rendered with default props 1`] = `
<Fragment>
  <Movie
    overview=""
    poster_path=""
    release_date=""
    runtime={0}
    tagline=""
    title=""
    vote_average={0}
  />
  <div
    className="search-description__wrapper"
  >
    <p
      className="movie-page__results-description"
    >
      Films by undefined genre
    </p>
  </div>
  <Connect(ResultsList) />
</Fragment>
`;

exports[`MoviePage component should be rendered with selectedMovie 1`] = `
<Fragment>
  <Movie
    genres={
      Array [
        "a",
      ]
    }
    overview=""
    poster_path=""
    release_date=""
    runtime={0}
    tagline=""
    title=""
    vote_average={0}
  />
  <div
    className="search-description__wrapper"
  >
    <p
      className="movie-page__results-description"
    >
      Films by a genre
    </p>
  </div>
  <Connect(ResultsList) />
</Fragment>
`;
